### ISSUE TITLE
Flow coming from closed nodes when using FlowDirectorDINF?

### ISSUE THREAD

**Issue by amanaster2** [landlab/landlab#811]
@kbarnhart (@saisiddu  told me to direct this to you!):
I was playing around with the different flow director methods, and I came across an oddity while using FlowDirectorDINF. It seems that flow is coming from closed nodes, which I didn't think should be happening... If, however, flow _can_ come from closed nodes, why don't I see that when using FlowDirectorD8? 
Below is the code I used to produce my grid and plots, along with two figures (FlowDirectorDINF vs. FlowDirectorD8).

Thank you!
Amanda

```
#%% Import packages
import numpy as np
import matplotlib.pyplot as plt

from landlab.components import FlowAccumulator, FlowDirectorDINF
from landlab.plot.imshow import RasterModelGrid
from landlab.plot.imshow import imshow_grid
from landlab.plot.drainage_plot import drainage_plot

#%% Create grid and plot
init = np.ones([100])
surface = (init*12) + np.random.rand(init.size)/100000.

mg = RasterModelGrid(10, 10, spacing=(1,1))
z = mg.add_field('topographic__elevation', surface + mg.node_y*0.05 + mg.node_x*0.05, at = 'node')

mg.set_closed_boundaries_at_grid_edges(True, True, True, True) 

plt.figure()
imshow_grid(mg,z)


#%% Get ordered nodes, drainage area
def ordered(grid, outlet_id=None):
    if outlet_id == None:
        outlet_id = np.argmin(grid.at_node['topographic__elevation'])
   
    grid.set_watershed_boundary_condition_outlet_id(outlet_id, 
                                                    grid.at_node['topographic__elevation'], 
                                                    nodata_value=-9999.)

    fa = FlowAccumulator(grid, flow_director = FlowDirectorD8)
    fa.run_one_step()
    (da, q) = fa.accumulate_flow()
    ordered_nodes = grid.at_node['flow__upstream_node_order']
    
    return ordered_nodes, grid, da, q

ordered_nodes, mg, da, q = ordered(mg)
plt.figure()
drainage_plot(mg, 'drainage_area')
```

![drainageplotdinf](https://user-images.githubusercontent.com/26778806/46898506-f051ff80-ce3e-11e8-8f7e-6805e7849a1c.png)
Using FlowDirectorDINF

![drainage_plotd8](https://user-images.githubusercontent.com/26778806/46898504-f051ff80-ce3e-11e8-983c-c6ffd7400567.png)
Using same code above but with FlowDirectorD8

**Comment by kbarnhart** [comment]
@amanaster2 thanks for pointing this out. Sorry for the delay, I was out of internet Fri-Sun.

I just looked into this and it is kind of a bug and kind of not. Let me elaborate.

1. You should consider the arrows in drainage plot to be useful for looking at the patterns of `flow__receiver_node` and `flow__receiver_proportions`. The arrows themselves do not provide any information about how much drainage area or discharge is being accumulated. That is why we are also plotting the drainage area as the tiled colors. 

You are correct that those yellow arrows leading from the closed nodes on the edge of the DINF plot shouldn't be there. This is the way in which what you have pointed out is a bug. Its super easy to fix (I did it on the bus this morning in about 2 minutes), but it is harder to write the tests that allow me to verify that I did it 100% correctly (e.g. because this needs to be a low priority for me it might be fixed in 2 days if I can do it on the bus, or it might be January).

At the same time that I fix this and write its tests, I'll also want to fix the issues pointed out in #667 (which has to do with DINF interacting correctly with closed nodes) and #802 (which has to do with route-to-multiple methods working correctly on completely flat surfaces). It turns out that implementing DINF by the original algorithm is not too hard to do, but doing that along side the boundary condition standards that we've developed in Landlab is pretty tricky (to do and to test). 

2. As to why this is not 100% a bug. Even though those yellow arrows exist, if you l look at the drainage area and discharge fields, you'll see that the closed nodes have zero drainage area and do not contribute to drainage area or discharge. This is because the drainage area of closed nodes gets zeroed out before it gets accumulated. 

Let me know if this doesn't make sense. 

So in summary, thank you so much for pointing this out (and making an example that made it super easy for me to reproduce your work) 👍 👍 👍 👍  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉  🎉 

If getting this to work in the near future is something you really need, please let me know and we can figure out a way to make it happen faster (it will probably involve you and me working together to develop good tests). 

If its not a problem to wait a bit, let me know too. 

Either way, on the next landlab telecom, I'll discuss whether we should throw up a quick warning in DINF to let people know that there are a few reasons for which it is not at 100% yet.

**Comment by amanaster2** [comment]
@kbarnhart Everything you explained makes perfect sense! I was trying out the different flow routing methods as a quick comparison -- I think I will be using the D8 router for the time being, so no rush on my side of things! Just wanted to bring it to your attention. Thank you, Katy!!
